New star catalog with Gaia DR3

henrysky opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

### Description
<!--- Please include a summary of the change and which issue is fixed. -->
<!--- Please also include relevant motivation and context. -->
<!--- List any dependencies that are required for this change. -->

Still draft and writing. See discussion in #3982

Major Changes:
- Star catalogs are built upon Hipparcos and Gaia DR3. Every stars brighter than magnitude ~6.5 should be included for visual purposes. In total having ~220 millions stars down to ~18 mag.
- New star catalogs format and codes to allow proper astrometry calculation. RA, DEC, PMRA, PMDEC, Parallax, Radial Velocity (and apperant magnitude) are calculated and displayed at epoch T for stars that have all required data.
- RA/DEC coordinates are now stored at ICRS epoch J2016 (so one does not need to propagate large amount of Gaia stars back to J2000 but Hippracos stars are forwarded to J2016), there is now epoch information at the header of each catalog so Stellarium can check if the epoch is the one expected. 
- Magnitude and color system remains the same as the old catalogs, meaning now V-band magnitudes for majority of the stars are estimated from Gaia G-band magnitude and G_bp-G_rp color with about ~0.03 mag scattering (see https://gea.esac.esa.int/archive/documentation/GDR3/Data_processing/chap_cu5pho/cu5pho_sec_photSystem/cu5pho_ssec_photRelations.html). B-V color are more challenging if one simply estimating from G-mag and G_bp-G_rp color as shown. As a result, the values in the new catalog are synethetic photometry computed from Gaia low resolution spectra (~0.03 mag scattering compared among Hipparcos stars).
- `Star1` in the catalog now has object types field and will be displayed among the field `Type` on top of existing information. Meaning of the display in its current state can be found at https://simbad.cds.unistra.fr/Pages/guide/otypes.htx.
- `Star1` that travel outside of their own zone in the far future/past are handled by pre-computing which stars will have this behavior and put them in a global zone (an extra zone for each geodesic level). So stars like Arturus can be displayed properly in the far future.
- Some stars being difficult to be selected are fixed.
- Search by Gaia source ID are added (both via SIMBAD and offline), so the star with that Gaia source ID will be selected instead of a custom object
- Updated documentation regarding the star catalogs and astrometry accuracy.

I will upload the code to generate the new catalog here: https://github.com/henrysky/stellarium_star_catalogs. Give me a few days to organize.

To-do list for Henry:
- [ ] Release source code

Fixes #18
Fixes #85
Fixes #221
Fixes #348
Fixes #362
Fixes #363
Fixes #364
Fixes #369
Fixes #398
Fixes #399
Fixes #677
Fixes #777
Fixes #1016
Fixes #1032
Fixes #1061
Fixes #1189
Fixes #1192
Fixes #1541
Fixes #1659
Fixes #1724
Fixes #2786
Fixes #3108
Fixes #3136
Fixes #3315

Many issues related to missing stars, false stars, incorrect information, astrometry inconsistent with other softwares...

### Screenshots (if appropriate):

### Type of change
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] This change requires a documentation update
- [x] Housekeeping

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

**Test Configuration**:
* Operating system: MacOS 15
* Graphics Card: Apple M3

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the [code style](http://stellarium.org/doc/head/codingStyle.html) of this project.
- [ ] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (header file)
- [x] I have updated the respective chapter in the Stellarium User Guide
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

alex-w commented:
@henrysky I've uploaded your catalogs into our GH page - https://github.com/Stellarium/stellarium-data/releases/tag/stars-3.0 - you may fix the URL's in JSON config for stars.

But (!) why this JSON config is not updating automatically anymore in users side?

alex-w commented:
Ouch... the `StarWrapper1::getObjectTypeI18n()` is should be updated too to re-enabling translation the types of stars

gzotti commented:
Before I start testing: Is it possible to have both catalogs installed and start one or the other branch without problems? (What about installations of several versions of Stellarium using the same user data dir?) I cannot risk messing up anything before later next week.

alex-w commented:
> @henrysky I've uploaded your catalogs into our GH page - https://github.com/Stellarium/stellarium-data/releases/tag/stars-3.0 - you may fix the URL's in JSON config for stars.

Damn! GH can't uploading file over 2GB, so, we should change URL for part 8 (https://dss.stellarium.org/stars-3.0/stars_8_2v0_2.cat) or change the range between magnitudes inside of parts and get a more parts for catalog.

alex-w commented:
> Before I start testing: Is it possible to have both catalogs installed and start one or the other branch without problems? (What about installations of several versions of Stellarium using the same user data dir?) I cannot risk messing up anything before later next week.

Yes, but automatic updating JSON data in users directory should be fixed for it

alex-w commented:
Checksum for part 8 in JSON file should be updated too

alex-w commented:
OK, this PR probably fix many related to star catalogs bugs and we should revised all of them: #18, #85, #221, #348, #362, #363, #364, #369, #398, #399, #677, #777, #1016, #1032, #1061, #1189, #1192, #1541, #1659, #1724, #2786

gzotti commented:
2GB! Note that some users may still use FAT32 (or other size-limiting FS?) for whatever reasons. File sizes should be kept under 2GB.

10110111 commented:
> Note that some users may still use FAT32

I doubt they'd use such large catalogs if they are limited by such file systems.

alex-w commented:
> I doubt they'd use such large catalogs if they are limited by such file systems.

Previously part 8 has size 0,5GB

gzotti commented:
This is part of my original concern about memory footprint. How much to load on which systems may depend on system capabilities. Some users may prefer to keep the current lightweight catalog. 4GB RAM is still very common. (Not sure about 1-2GB apart from the RPi platform and maybe other SBCs usable as telescope computers.)

henrysky commented:
> @henrysky I've uploaded your catalogs into our GH page - https://github.com/Stellarium/stellarium-data/releases/tag/stars-3.0 - you may fix the URL's in JSON config for stars.

Thanks! I have updated the JSON config for stars

> [edit] Please fix line `static const int StarCatalogFormatVersion = 12;` in file `StarMgr.cpp`

Fixed in the branch

> Ouch... the StarWrapper1::getObjectTypeI18n() is should be updated too to re-enabling translation the types of stars

Fixed in the branch but the SIMBAD 3-letters object type code covering stars, set of stars, quasar, galaxies.... probably need to convert to words (can use the one provided on SIMBAD too). Parts of the logic inside ``StarWrapper1::getObjectType`` can also merged to this object type column in the `Star1` catalog. Any opinion?

> Damn! GH can't uploading file over 2GB, so, we should change URL for part 8 (https://dss.stellarium.org/stars-3.0/stars_8_2v0_2.cat) or change the range between magnitudes inside of parts and get a more parts for catalog.

I think I will break the >15 mag catalogs into slightly smaller magnitude range to make it more flexible to donwload/storing them/

> This is part of my original concern about memory footprint. How much to load on which systems may depend on system capabilities. 

I don't think stellarium uses much more memory with the new star catalog compared to 24.3 both with all star catalogs at least on my computer.

Atque commented:
I see a few stars lack proper motion and distance (Caph, Navi and Phecda for instance). Possibly more.

I guess there was an error compiling the catalog?

alex-w commented on stars/default/defaultStarsConfig.json:
```diff
@@ -86,8 +86,8 @@
 			"count": 138.4,
 			"magRange": [16.5, 18],
 			"sizeMb": 0,
```
Please fix size of file

alex-w commented:
> I think I will break the >15 mag catalogs into slightly smaller magnitude range to make it more flexible to donwload/storing them.

Good solution IMHO

alex-w commented:
> Fixed in the branch but the SIMBAD 3-letters object type code covering stars, set of stars, quasar, galaxies.... probably need to convert to words (can use the one provided on SIMBAD too). Parts of the logic inside StarWrapper1::getObjectType can also merged to this object type column in the Star1 catalog. Any opinion?

I've added cosmetic fix for star types. Adding support the SIMBAD's 3-letters object type codes for each object is not possible at the moment without many refactoring. Adding "humar readable" explanation is possible, but not need at the moment IMHO (see morphological description for DSO objects).

alex-w commented:
> I see a few stars lack proper motion and distance (Caph, Navi and Phecda for instance). Possibly more.

Absolute magnitudes for stars also seems very weird - for example alpha Cassiopeia has -12.02 now

henrysky commented:
> I see a few stars lack proper motion and distance (Caph, Navi and Phecda for instance). Possibly more.

Ah, good catch. There is some logic flaws in the cross-matching between Hipparcos and Gaia. All stars dim enough to be observed by Gaia but bright enough to have no Gaia pmra/pmdex/parallax will have this issue. I am figuring out a solution.

> Absolute magnitudes of stars also has very weird values - for example alpha Cassiopeia has -12.02 now

The parallax used to compute absolute magnitude is 100 times too small (but the parallax is correct in the catalog and correctly displayed). I will fix it now.

alex-w commented:
Ouch... please check spectral classes of stars also

Atque commented:
> Just started a first build. Very promising! However, see e Eri (82 Eri, HIP15510). Going backwards around November -30070 (yes, I know, very early. Just an arbitrary example, I have seen a few similar ones) the star vanishes. I have not traced a reason. In the current Zone catalog, a star may leave its zone triangle and vanish, maybe it's the same phenomenon.

I have seen this phenomenon in many other software. It seems difficult to solve.

sunshuwei commented:
Very much looking forward to Gaia replacing Hipparcos.
Will other catalog (HD, HR, SAO) numbers be added to the stars not covered by the Hipparcos catalog in the future? Gaia basically covers these stars.
And with the Gaia star chart, can sky culture use non Hipparcos stars?

alex-w commented:
> Very much looking forward to Gaia replacing Hipparcos. 

Gaia not replacing Hipparcos yet, only NOMAD and partially Tycho 2.

alex-w commented:
> Ouch... please check spectral classes of stars also

Yes, most stars has wrong spectral classes

alex-w commented:
I see many stars from Gaia catalog with distance, but without absolute magnitudes - why?

alex-w commented:
> I simply forget to add absolute magnitudes display to Star2 because parallax is a newly added information to them. (edit: actually I've thought of adding Absolute Magnitude as commented in the code, I was thinking about extinction because many of them might have high extinction so will mess up absolute magnitude (because apperant magnitude will be lower). If you think absolute magnitude is useful sure I can add them too)

Yes, absolute magnitude for stars with parallax/distance is important (at least for didactic purposes)

henrysky commented:
> Yes, most stars has wrong spectral classes

Spectral class of stars being wrong should be fixed. There is a slight reduction in number of stars in the first few catalogs because now HIP stars and their companions are better matched with Gaia DR3 counterpart so much less stars are drawn on top of each other (because they are being the same star).

> Yes, absolute magnitude for stars with parallax/distance is important (at least for didactic purposes)

Absolute magnitudes and distances are added now only if less than 20% parallax uncertainty. But parallax are displayed as long as there are measurements.

> Hopefully the numebr of stars in this case in small and won't bring much performance impact. I can quickly later to give you a number estimate.

I've estimated those numbers, going back to -100000 and going forward to 100000. ~13% of stars will leave their own zone in star catalog 0,  14.6% of stars will leave their own zone in star catalog 1 and 22.8% of stars will leave their own zone in star catalog 2.

> I think this is the reason for this version of star catalogs too. One solution I am thinking of is to create a level -1 Zone (the whole sky is a single zone) such that we put those HIP stars in the level -1 zone...

I think splitting the large star catalogs of the same geodesic grid level into a few smaller files and having a "-1 zone" require changes in quite a few places, which at this rate might be worth to switch from geodesic grid to HEALPix which will allow efficient searching of a specific Gaia ID because each Gaia ID are encoded with the info for HEALPix index of level 12 and lower (i.e., given a gaia ID, you know which zone they are going to be in at a certain level<12). But we will be looking at a timeline of a few months probably.

alex-w commented:
> Spectral class of stars being wrong should be fixed. There is a slight reduction in number of stars in the first few catalogs because now HIP stars and their companions are better matched with Gaia DR3 counterpart so much less stars are drawn on top of each other (because they are being the same star).

But files are changed and checksums should be updated in JSON file too to avoid problems.

> Absolute magnitudes and distances are added now only if less than 20% parallax uncertainty. But parallax are displayed as long as there are measurements.

OK, good solution!

alex-w commented:
@henrysky I see in the log now:
```
Loading star catalog: /opt/stellarium/share/stellarium/stars/default/stars_3_1v0_5.cat - error - file format is bad.
```

henrysky commented:
Actually the performance is not that bad if every stars in the first 3 catalogs are checked if they are inside the screen (instead of by zone). The brute-force check is turned on if it is more than 3000 years away from 2016. You guys can check the latest commit (just a temporary solution, the code/logic sure can be better).

Atque commented:
> Actually the performance is not bad if every stars in the first 3 catalogs are checked if they are inside the screen (instead of by zone). The brute-force check is turned on if it is more than 3000 years away from 2016. You guys can check the latest commit (just a temporary solution without modifying the star catalogs, the code/logic sure can be better). Stars like Arcturus/Barnard's star can now be properly displayed in the future/past even when zoomed in.

This commit keeps stars selectable in the distant future or past too.

I see a few double stars only show PM for one component, and the other (bearing only a Gaia designation) being completely stationary (Porrima, Sabik).

gzotti commented:
On stars leaving their Zone: Are there stars migrating by more than one adjacent zone? It may be enough to just check all visible zones plus their immediate neighbors. (Or expand "neighbors" further if this is not enough.)

henrysky commented:
Hi all, I think I have found the root cause of #3832 and #3733 and have pushed a commit to fix that issue in this branch. The root cause is abbrevation is incorrectly apply when searching solar system object leading to abbrevation needed to be applied to `cleverfind()` function which leads to a wrong mouse click pixel location on screen which lead to some stars are unselectable (especially in those sky region most affected by abbrevation). I believe I have fixed the root cause and you can select easily and correctly for plutos and all its moons as well as stars k Cru cluster star.

On the other hand, I have added functionality to allow search by Gaia ID (via both SIMBAD and offline). So if a particular stars with a Gaia ID in the star catalog, that star will be selected instead of a custom object.

I have also added a global zone at each geodesic level for those stars that move more than 1 zone in -110000 to 110000 instead of brute-forcing. @gzotti you are right that only stars moved more than 1 zone need to be taken care of because border zones are checked but not border of border zones. Number of stars that are going to move more than 2 zones are much smaller (less than 1% of the stars in the first 3 catalogs)

alex-w commented:
Hmm... a quick test and I got a crash dumps in linux (revison [d8a7bc7](https://github.com/Stellarium/stellarium/pull/3992/commits/d8a7bc751f71280ac2aff878f2b46a2ebbab5ffd))

henrysky commented:
How about with the latest [8b91971](https://github.com/Stellarium/stellarium/pull/3992/commits/8b91971596038a04a1a6d1cb74a52be13f24ee5a)? Works on my MacOS. Other star catalogs (those not included in stellar out of the box) are not going to work because of an exra zone added as a global zone for each geodesic level, I am working on those currently.

alex-w commented:
No crash anymore!

alex-w commented on stars/default/defaultStarsConfig.json:
```diff
@@ -1,93 +1,93 @@
 {
-	"version": 13,
+	"version": 14,
```
Did you changed the version number in C++ code?

henrysky replied:
I just recognized this and force pushed to bump the version.

alex-w replied:
Hint: `static const int StarCatalogFormatVersion`

henrysky replied:
I will change that in the morning or you can push the change😅 i have turned off my laptop to go to sleep. I will make a checklist for myself to keep track of what action need to be taken each time I update catalogs and JSON.

alex-w commented:
Hmm... The search tool cannot found Gaia stars.

alex-w commented:
I doing not all checks and tests yet, but the pull request is mature almost - of course related changes in the User Guide are needed. I think we may unmark “draft” for PR.

Probably the PR may be targeted to release 24.4.

10110111 commented:
> Probably the PR may be targeted to release 24.4.

To the nearest release? No! This PR has diff stats of +8395/-243319 lines, but hasn't been tested by anyone but a few people. Its merge shouldn't happen before the nearest release.

gzotti commented:
I just see a crazy flashing: Center Arcturus, date -84796-4-30 (not sure how much it matters), fov 55-65. Else happy it is still visible!
Yes, it's too early to enable this, but it looks promising. Needs more testing and working out the side effects. In far past/future framerate seems to go down considerably, etc.

alex-w commented:
@10110111 you're right - just producing the binary packages for this PR should be enough for wide testing by community.

OK, the target milestone is 25.1 now

alex-w commented:
> I just see a crazy flashing: Center Arcturus, date -84796-4-30 (not sure how much it matters), fov 55-65. Else happy it is still visible! 

Yes, I can see flashing too - and I see the very fast movement the 39 Ser star at flashing moment

gzotti commented:
Against which "reference solution" can we actually test the new proper motion? Is there anything besides CdC? How accurate is orbitsimulator.com?

Atque commented:
> Against which "reference solution" can we actually test the new proper motion? Is there anything besides CdC? How accurate is orbitsimulator.com?

AFAIK Orbitsimulator is very good, but sadly its coordinate system is simplified, so only relative positions between stars can be used.

henrysky commented:
I have updated the star catalogs to solve various issues reported by you guys so far. Extra catalogs are the same so no need to be updated so far.

> Hmm... The search tool cannot found Gaia stars.

@alex-w Searching Gaia stars is working fine for me. Are you sure you have input a valid Gaia source ID (Gaia DR3 xxxxxxx....)? Like `Gaia DR3 4472832130942575872`.

> > I just see a crazy flashing: Center Arcturus, date -84796-4-30 (not sure how much it matters), fov 55-65. Else happy it is still visible!
> 
> Yes, I can see flashing too - and I see the very fast movement of the 39 Ser star at flashing moment

@gzotti @alex-w I cannot reproduce this issue but I did found the root cause for the problem of 39 Ser. 39 Ser is a star in a 6.0-7.5 mag star catalog where it gets brighter than catalog minimum magnitude in the past and cause issue by looking up drawing paramaters outside of `rcmag_table` range. The issue has been fixed by constructed `rcmag_table` brighter than each star catalog minimum magnitude (HIP 16757 actually get way brighter in the past). Can you check if the latest commit has fixed the problem? If not can you attach a video of the issue?

> Needs more testing and working out the side effects. In far past/future framerate seems to go down considerably, etc.

@gzotti Low framerate in far past/future issue might have caused by recalculating magnitude and then check EVERY single stars in catalog 0/1/2 to make sure stars that get much brighter in the past/future get drawn instead of not surviving magnitude cutoff. Now stars that get much brighter (delta mag is <-0.3, meaning the stars can get more than 0.3 mag bright at some point) in the past/future are put in the global zone in each geodesic level. So only a fraction of the stars are checked (i.e., those in the global zone) and make sure they are drawn in the past/future even zoomed out but not those stars in normal zones. Stars that get much dimmer in the past/future stay in their zone because the issue is bright stars potentially not get drawn but not dim stars get drawn (they will still be drawn according to their recomputed magnitude at that year NOT catalog value if information avaliable).

You can see if the latest commit run smoother in far past/future.

alex-w commented:
The flashing is gone in latest commit

alex-w commented:
> Searching Gaia stars is working fine for me. Are you sure you have input a valid Gaia source ID (Gaia DR3 xxxxxxx....)? Like `Gaia DR3 4472832130942575872`.

Please see screenshots:
![stellarium-023](https://github.com/user-attachments/assets/3ae925f6-6bef-496e-831e-782fa736fb78)
![stellarium-024](https://github.com/user-attachments/assets/ca3d39e4-8ddc-4948-a7ee-98a3bab623d2)

So, probably, `getGaia()` gives wrong identifiers

henrysky commented:
> So, probably, `getGaia()` gives wrong identifiers

I have just fixed the issue because getGaia() has a return type of `long` which has no guranttee the integer is 64bit, now I've switched to `int64_t` and should have fixed the issue.

alex-w commented:
> Against which "reference solution" can we actually test the new proper motion?

Not a reference solution, but very good start point for tests - https://articles.adsabs.harvard.edu/pdf/1994QJRAS..35....1M (the link from issue #353)

alex-w commented:
> > So, probably, `getGaia()` gives wrong identifiers
> 
> I have just fixed the issue because getGaia() has a return type of `long` which has no guranttee the integer is 64bit, now I've switched to `int64_t` and should have fixed the issue.

This is strange, but when I select any Gaia stars in search tool I got a selected Rigel

henrysky commented:
> This is strange, but when I select any Gaia stars in search tool I got a selected Rigel

Searching Gaia stars get you to Rigel? That is weird, Rigel does not even has a Gaia source id for being too bright!

> Against which "reference solution" can we actually test the new proper motion? Is there anything besides CdC? How accurate is orbitsimulator.com?

I think the "reference solution" can be IAU SOFA library: http://www.iausofa.org/ with astrometry related tool documentation: http://www.iausofa.org/2023_1011_C/sofa/sofa_ast_c.pdf but I have never tried it. `astropy` is a good alternative but I have never been in a situation before where very precise RA/DEC are needed so need to see how good they are for this task but I would assume `astropy` is good since I often use it in my research for other purposes.

gzotti commented:
It would be great to set up a few representative stars (any random mag10-12 stars, and a few outstanding "runners" [Barnard, ...]) into an automated unit test. Work out reference data with e.g. astropy, and see how close your solution comes to that, in J2025 and say e.g. &pm;5000 and &pm;10.000 years. (Or even more? What are the guarantees for astropy regarding long-time simulation?)

alex-w commented on stars/default/stars_hip_cids_0v0_0.cat:
Please update JSON also

henrysky commented:
I have updated the star catalogs again. A main update is I've put stars that are brighter than 6.5 mag who will cross at least ONE zone in -100,000 to 100,000 years into the global zone (instead of crossing at least two zone). I've found the level 0 triangles are so big that you can zoom enough to NOT have border triangles checked so some stars still disappearing in the future and in the past. A minor update is I have put ``Ross 248`` into the third star catalog even thought it is very dim now because it is an interesting object which will be the closest star to us in ~40,000 years. **If you think there are interesting objects deserving full 6D astrometry propagation that are dimmer than magnitude 10.5 and does not have a HIP numbers, you can comment below with its Gaia source ID so I can put them in the bright star catalogs.** For dim star catalogs, only level 5 catalog is changed and is here: https://www.astro.utoronto.ca/~hleung/shared/stellarium_new_catalog/trial4/stars_5_1v0_4.cat

I have also added astrometry test on 2 stars at +/- 5,000 years as well as +/- 50,000 years compared to astropy, asserting deviation in those epoch is within 36 arcseconds for RA/DEC, 1 mas for parallax, 5 mas/yr for PMRA/PMDEC, and 2km/h for radial velocity. The main issue is not the algorithm in stellarium is wrong or we are too simple to not take relativistic effects/light traveling time into account, but small numerical error for very high proper motion stars like Barnard's star actually is signaficant like 50,000 years from now. So I think this is as good as we can do in terms of limited computational resources but Don't expect <1 arcsecond precision 50,000 years from now.

I have also started to update the docs. Let me know if there are other places I need to update besides the two I am working on (Appendix C star catalogs and Appendix F astrometry accuracy).

gzotti commented:
36 arcsecs (or even an arcmin) difference in 50ky sounds perfectly adequate to me. We just should avoid real systematic errors like it may have been before. 
My main interest is the naked-eye sky (Arcturus &friends not vanishing by Zone error, &alpha; Cen not flying apart, ...), I cannot name favourite mag 10 runners without own investigation. Will Ross 248 be visible to the naked-eye then?

henrysky commented:
I Don't think we have systematic errors on RA/DEC anymore. ``Ross 248`` is not gonna be naked eyes visible (maxed at ~9.8 mag) but it is going to be our closest stars for about 9,000 years covered by stellarium date range (https://en.wikipedia.org/wiki/Ross_248), something maybe of interest for educational purposes to show how close it can get.

I am also marking this PR ready for review unless there is other existing issue that can potentialy be fixed with changing the catalog format. I do not have any more features to add and I think I have fixed bugs reported here so far.

Atque commented:
We ought to make sure all Bayer and Flamsteed designations are added in the new catalog. For example, 41 Ari is only refered to as its Gaia designation (and lacks PM).

![bild](https://github.com/user-attachments/assets/5520dbb9-69b9-4889-87fa-d46cc2bb485d)

alex-w commented:
> We ought to make sure all Bayer and Flamsteed designations are present in the new catalog. For example, 41 Ari is only refered to as its Gaia designation (and lacks PM).

It's a task for next step (in my todo already :) ), not for current branch

henrysky commented:
> We ought to make sure all Bayer and Flamsteed designations are present in the new catalog. For example, 41 Ari is only refered to as its Gaia designation (and lacks PM).

41 Ari was cross-matched to a wrong star near the true 41 Ari (if you search 41 Ari, it will point it to another stars with HIP 13209A). I will fix it later to see if we have more bright stars get matched incorrectly.

alex-w commented on src/core/modules/Star.hpp:
```diff
@@ -178,7 +178,7 @@ struct Star
                                  float    dyrs) const
    {
-      if (getX2() != 0. || getDx2() != 0.) {
+      if (getPreciseAstrometricFlag()) {
```
Unit test is fails now

henrysky replied:
🥲 I did not think it throught and that is actually a wrong logic. Have reverted the commit back. But still need to solve the issue if a stars is `Star1` having no parallax but with radial velocity in the catalog will result in nan/inf in info string.

alex-w reviewed (approved):
Thank you very much!

Atque commented:
I've been testing, and two stars still have stationary doubles: Xi UMa (Alula Australis) and Porrima. Otherwise, this looks finished to me.

gzotti commented on guide/app_star_catalogue.tex:
```diff
@@ -45,44 +44,41 @@ \subsection{General Description}%\label{general-description}
 \begin{description}
-\item[\file{stars\_hip\_cids\_0v0\_0.cat}]
-\item[\file{stars\_hip\_sp\_0v0\_3.cat}]
+\item[\file{object\_types\_v0_1.cat}]
```
backslash missing in front of third underscore.

gzotti commented on guide/app_star_catalogue.tex:
```diff
@@ -209,32 +204,50 @@ \subsubsection{Star Data Records}%\label{star-data-records}
-mag            & 13 & unsigned char      & 1 & magnitude level in the V-I colour. This value refers to one of 256 discrete steps 
-                                               in the magnitude range for the file\\%\midrule
-sp\_int        & 14 & unsigned short int & 2 & index into an array of spectral type descriptions 
+gaia_id        &  0 & long long          & 8 & Gaia source id\\%\midrule
```
must be gaia\_id for typesetting

gzotti commented on guide/app_star_catalogue.tex:
```diff
@@ -209,32 +204,50 @@ \subsubsection{Star Data Records}%\label{star-data-records}
+dx0            & 20 & int                & 4 & First component of the star's proper motion vector at the catalog epoch scaled by $1\times10^3$\\%\midrule
+dx1            & 24 & int                & 4 & Second component of the star's proper motion vector at the catalog epoch scaled by $1\times10^3$\\%\midrule
+dx2            & 28 & int                & 4 & Third component of the star's proper motion vector at the catalog epoch scaled by $1\times10^3$\\%\midrule
+b\_v           & 32 & short              & 2 & $B-V$ colour in milli-mag unit.
```
This and the next table line must have \\\\ as linebreak marker.

gzotti commented on guide/app_star_catalogue.tex:
```diff
@@ -209,32 +204,50 @@ \subsubsection{Star Data Records}%\label{star-data-records}
+b\_v           & 32 & short              & 2 & $B-V$ colour in milli-mag unit.
+mag            & 34 & short              & 2 & $V$-band magnitude in milli-mag unit.
+plx            & 36 & unsigned short     & 2 & parallax of the star in 0.02 mas unit.\\
+plx_err        & 38 & unsigned short     & 2 & parallax error of the star in 0.01 mas unit.\\
```
plx\\_err

gzotti commented on guide/app_star_catalogue.tex:
```diff
@@ -209,32 +204,50 @@ \subsubsection{Star Data Records}%\label{star-data-records}
-dx0            & 16 & int                & 4 & proper motion of the star in axis 1\\%\midrule
-dx1            & 20 & int                & 4 & proper motion of the star in axis 2\\%\midrule
-plx            & 24 & int                & 4 & parallax of the star. To get the actual value, divide by 10000.\\\bottomrule
+obtype		   & 44 & unsigned char      & 1 & object type classification index. The index corresponds to the line number in the file \file{object\_types_v*.cat} - 1.\\
```
> \file{object\\_types_v*.cat}

\file{object\\_types\\_v*.cat}

gzotti commented on guide/app_star_catalogue.tex:
```diff
@@ -243,14 +256,15 @@ \subsubsection{Star Data Records}%\label{star-data-records}
-                                          in the magnitude range for the file\\%\midrule
-mag  & 75 bits & unsigned int &  5 bits & magnitude level in the V-I colour. This value refers to one of 32 discrete steps 
-                                          in the magnitude range for the file\\\bottomrule
+gaia_id        &  0 & long long          & 8 & Gaia source id\\%\midrule
```
same as above, gaia\\_id and int64

gzotti commented on guide/app_star_catalogue.tex:
```diff
@@ -209,32 +204,50 @@ \subsubsection{Star Data Records}%\label{star-data-records}
+where $\mu_{\alpha^*}\equiv\mu_{\alpha}\cos{(\delta)}$.
 
 \begin{table}[p]
 \begin{tabular}{llllp{55mm}}\toprule
```
After these changes, the paragraph width could be enlarged to 75mm. (55mm-->75mm)

gzotti commented on guide/app_star_catalogue.tex:
```diff
@@ -243,14 +256,15 @@ \subsubsection{Star Data Records}%\label{star-data-records}
+gaia_id        &  0 & long long          & 8 & Gaia source id\\%\midrule
+x0             &  8 & int                & 4 & Right ascension $\alpha$ of the star in mas unit.\\%\midrule
+x1             & 12 & int                & 4 & Declination $\delta$ of the star in mas unit.\\%\midrule
+dx0            & 16 & int                & 4 & Proper motion $\mu_{\alpha}$ of the star in $mu$as/year%\midrule
```
This and next 3 lines: again, please use \\\\ as line end marker (The "%\midrule" is a testwise commented-away previous horizontal line and can go away. )

Also, this and next line, use "$\mu$as/year"  (with backslash) to typeset "&mu;as/year"

gzotti commented on guide/app_star_catalogue.tex:
```diff
@@ -243,14 +256,15 @@ \subsubsection{Star Data Records}%\label{star-data-records}
+b\_v           & 24 & short              & 2 & $B-V$ colour in milli-mag unit.
+mag            & 26 & short              & 2 & $V$-band magnitude in milli-mag unit.
+plx            & 28 & unsigned short     & 2 & parallax of the star in 0.01 mas unit.\\
+plx_err        & 30 & unsigned short     & 2 & parallax error of the star in 0.01 mas unit.\\
```
again plx\\_err. In the last line you can omit the trailing \\\\.

gzotti commented on guide/app_star_catalogue.tex:
```diff
@@ -259,12 +273,11 @@ \subsubsection{Star Data Records}%\label{star-data-records}
-                                          in the magnitude range for the file\\%\midrule
-mag  & 43 bits & unsigned int &  5 bits & magnitude level in the V-I colour. This value refers to one of 32 discrete steps 
-                                          in the magnitude range for the file\\\bottomrule
+gaia_id        &  0 & long long          & 8 & Gaia source id\\%\midrule
```
gaia\\_id

gzotti commented on guide/app_star_catalogue.tex:
```diff
@@ -259,12 +273,11 @@ \subsubsection{Star Data Records}%\label{star-data-records}
+gaia_id        &  0 & long long          & 8 & Gaia source id\\%\midrule
+x0             &  8 & int                & 3 & Right ascension $\alpha$ of the star in 0.1 arcsecond unit.\\%\midrule
+x1             & 11 & int                & 3 & Declination $\delta$ of the star in 0.1 arcsecond unit.\\%\midrule
+b\_v           & 14 & unsigned short     & 1 & $B-V$ colour in 25 milli-mag unit offset by $+1$ mag.
```
add \\\\ to end the table line.

gzotti commented on src/core/StelObject.hpp:
```diff
@@ -332,7 +332,8 @@ public slots:
 	//! @param flags
 	//! @param decimals significant digits after the comma.
-	virtual QString getMagnitudeInfoString(const StelCore *core, const InfoStringGroup& flags, const int decimals=1) const;
+	//! @param magOffset magnitude offset to apply (can be due to changing distance)
```
Can you please describe this new parameter in more detail? Or maybe it is described elsewhere?

alex-w replied:
OK, this is final point who need to fixation

henrysky replied:
I have described more. Basically when the distance of stars change because of proper motion/radial motion, magOffset is used to compute new apparant magnitude.

gzotti commented on src/core/modules/Star.hpp:
```diff
@@ -22,266 +22,394 @@
 static inline float IndexToBV(int bV)
 {
-	return static_cast<float>(bV)*(4.f/127.f)-0.5f;
+   return static_cast<float>(bV) * (4.f / 127.f) - 0.5f;
```
indentation looks broken

gzotti commented:
Actually for @alex-w: I haven't done so in a long time now. After loading a new star catalog it appears the stars are immediately used. At the end of catalog downloads however, the message is something like "All catalogs downloaded, restart Stellarium to load them". I think this last sentence could be removed, right?

gzotti commented:
Just running on RPi4. Mem footprint is very relaxed (comparable to the old). I don't think I had ever downloaded all star catalogs. Now I start seeing some loss in framerate at level 7 of 9. 12fps is still acceptable, though. I dare going still deeper...
What is the recommended way to remove a star catalog? Or could we just deactivate somehow? I guess only manually, editing starsConfig.json?

And important for comparisons against older versions or if there is just other needs to run older versions: Is it possible to keep all catalogs under one Stellarium user dir? The star catalog's name is always "default" and has been probably since 2006. If there is need for separate catalogs, this new one could be named e.g. hip_gaia, so that only somewhere in code (StarMgr.cpp) the directory name is adjusted from `default` to `hip_gaia`. Note that I don't say we must allow loading the old catalog into new program versions by config, but this is such a significant change that we should think about such mixed legacy installations.

alex-w commented:
> Just running on RPi4. Mem footprint is very relaxed (comparable to the old). I don't think I had ever downloaded all star catalogs. Now I start seeing some loss in framerate at level 7 of 9. 12fps is still acceptable, though. I dare going still deeper... What is the recommended way to remove a star catalog? Or could we just deactivate somehow? I guess only manually, editing starsConfig.json?

Just remove specific *.cat files - no need manually editing json file.

> And important for comparisons against older versions or if there is just other needs to run older versions: Is it possible to keep all catalogs under one Stellarium user dir? The star catalog's name is always "default" and has been probably since 2006. If there is need for separate catalogs, this new one could be named e.g. hip_gaia, so that only somewhere in code (StarMgr.cpp) the directory name is adjusted from `default` to `hip_gaia`. Note that I don't say we must allow loading the old catalog into new program versions by config, but this is such a significant change that we should think about such mixed legacy installations.

JSON file will be recreated if their version number is not equal to expected one in the binary package - as result you may use one place to storage all possible catalogs ;)

alex-w commented:
> Actually for @alex-w: I haven't done so in a long time now. After loading a new star catalog it appears the stars are immediately used. At the end of catalog downloads however, the message is something like "All catalogs downloaded, restart Stellarium to load them". I think this last sentence could be removed, right?

Maybe, I just don’t test loading catalogs on-the-fly…

gzotti commented:
OK, but running e.g. 25.1, 1.2 and 0.15.2 on the same PC (for comparison, linked with RemoteSync or for whatever reasons; or alternating between versions for with/without ASCOM stuff), sharing the same STEL_USERDIR, may trigger conflicts when the second launched overwrites the first's JSON. (And does this mechanism work for old versions like 0.15?) This looks like dangerous design for me, and goes against whatever may have been planned in 2010 when "default" was apparently introduced. @xalioth, @10110111 ?

alex-w commented:
> OK, but running e.g. 25.1, 1.2 and 0.15.2 on the same PC (for comparison, linked with RemoteSync or for whatever reasons; or alternating between versions for with/without ASCOM stuff), sharing the same STEL_USERDIR, may trigger conflicts when the second launched overwrites the first's JSON. (And does this mechanism work for old versions like 0.15?)

Overwriting is possible after comparing versions in JSON file and in binary package, plus our catalogs is not changed frequently…

> This looks like dangerous design for me, and goes against whatever may have been planned in 2010 when "default" was apparently introduced. @xalioth, @10110111 ?

Well, the `default` section for our catalogs (stars + dso) and some custom catalogs was ambitious idea, but it never implemented and the code for switching between various collections was removed a long time ago…

gzotti commented:
OK, no need to add switching code. Does a simple change of the catalog path in StarMgr::copyDefaultConfigFile() and StarMgr::init() lead to the suggested separation of "all versions which used the HIP/TYC/NOMAD catalog (named "default" from 2010 to 2024)" and "versions starting with 24.4+ using this new HIP/GAIA catalog". Old versions keep working with their current logic (in theory even catalog bugs may still be fixed, as unlikely as this is), while new versions will just never have the necessity to touch the "default" catalog. If you have no old version installed, the name of the (only) catalog will just be hip_gaia (or whatever we name this now, "v25", just not "default"). Truly incompatible changes should be handled like this, IMHO.

alex-w commented:
> OK, no need to add switching code. Does a simple change of the catalog path in StarMgr::copyDefaultConfigFile() and StarMgr::init() lead to the suggested separation of "all versions which used the HIP/TYC/NOMAD catalog (named "default" from 2010 to 2024)" and "versions starting with 24.4+ using this new HIP/GAIA catalog". Old versions keep working with their current logic (in theory even catalog bugs may still be fixed, as unlikely as this is), while new versions will just never have the necessity to touch the "default" catalog. If you have no old version installed, the name of the (only) catalog will just be hip_gaia (or whatever we name this now, "v25", just not "default"). Truly incompatible changes should be handled like this, IMHO.

Maybe you're right and string `stars/default` in the code (C++ & CMake) should be changed to `stars/hip_gaia`

alex-w commented:
> Actually for @alex-w: I haven't done so in a long time now. After loading a new star catalog it appears the stars are immediately used. At the end of catalog downloads however, the message is something like "All catalogs downloaded, restart Stellarium to load them". I think this last sentence could be removed, right?

Just checked right now - no, we shouldn't remove this last sentence, because stars are not loaded from catalogs on-the-fly (at the moment of loading catalog)

gzotti commented:
RPi3 also works, at least with extra cats stars_4 and stars_5. It downloads stars_6 but initialisation fails due to lack of memory. I did not realize this and happily started downloading even stars_7, then stellarium crashed with std::bad_alloc. But I think RPi3 is just the minimum platform to be used for limited installations, so going to about Pluto magnitude (13.75) should be fine.

henrysky commented:
I am going to fix issue of the two stars brought up by @Atque. @alex-w Should I bump all the star catalogs version to 1.0? Beacuse the new star catalogs are not compatible with the old v0 star catalogs?

alex-w commented:
> @alex-w Should I bump all the star catalogs version to 1.0? Beacuse the new star catalogs are not compatible with the old v0 star catalogs?

I see no necessary repacking catalogs

henrysky commented:
I have fixed all missing pm/parallax issue mentioned by others. @alex-w Only stars4/5 is changed and is here https://www.astro.utoronto.ca/~hleung/shared/stellarium_new_catalog/trial5/.

alex-w commented on src/core/StelObject.hpp:
```diff
@@ -332,7 +332,7 @@ public slots:
 	//! @param flags
 	//! @param decimals significant digits after the comma.
-	//! @param magOffset magnitude offset to apply (can be due to changing distance)
+	//! @param magOffset magnitude offset to apply to display final apperant magnitude, are used if a star distance has changed in the past/future
```
Maybe _apparent_?

henrysky replied:
done!

alex-w commented:
> I have fixed all missing pm/parallax issue mentioned by others. @alex-w Only stars4/5 is changed and is here https://www.astro.utoronto.ca/~hleung/shared/stellarium_new_catalog/trial5/.

OK, done!

alex-w commented:
@gzotti @10110111 any comments?

P.S. A good question - squash or rebase?

10110111 commented:
Some new files here lack the POSIXly-desired trailing newlines, which is indicated by git-diff and by GitHub's diff viewer. These should be added.

> P.S. A good question - squash or rebase?

Are there any testable versions among these commits, so that it would make sense to run, say, git-bisect on them? If yes, I suppose it's better to rebase.

In any case, the changes to documentation (the Guide first of all) should be separated and squashed into a single commit after all the others.

gzotti commented:
Sorry our ISP was offline for some time in mid-afternoon for whatever reason. Dec.24 overload? (about 15.000 reports at my ISP.)

 The current incremented version number is OK for me. Just the new path was important to allow parallel installations, thanks. 
I just wanted to follow the morphing of our 3-star-linked constellation artwork over the millennia. However it seems the projection is computed only at loading of the respective skyculture. E.g. loading Inuit around 85000AD looks weird. But this is not for this PR. Now SC authors face a new challenge: define the 3-star link with low-PM stars :-)  (Or make distortion part of the show...)

I think there is no half-working in-between version to ever bisect. 24.4 has the HIP/TYC/NOMAD catalog, and this PR switches to HIP/GAIA DR3. For me a rebase&squash would be OK, even if the changes is large. Also, the directory issue may break any bisects.  Of course, later fixes will be separate.  So, for me it's OK.

gzotti reviewed (approved):
Cosmetics like the empty line at end should be fixed, I second that (whatever git or CodeFactor report). 

Else, this looks OK for me, many thanks!

gzotti commented:
I just see that TestAstrometry fails, at least on Windows. RA stays at 0 after searchGaiaIDepochPos(,. ,. RA, ...) @henrysky ?

henrysky commented:
> I just see that TestAstrometry fails, at least on Windows. RA stays at 0 after searchGaiaIDepochPos(,. ,. RA, ...) @henrysky ?

Fixed in fecb74827cacc0cd12d649f2943df704d117589d, issue is how the testAstrometry finds the star catalog for testing depending where you launch the test, nothing is wrong with the main code that testAstrometry is designed to test.

gzotti commented:
Still no luck. I build in QtCreator and run the tests just there, so no installation. 

Fail: TestAstrometry::initTestCase
'!files.isEmpty' returned FALSE. (No matching star catalog found at D:/)

My star catalogs from master are in D:\StelDev\GIT\stellarium\stars\hip_gaia3. 
Build directory is D:\StelDev\GIT\build-stellarium-Desktop_Qt_6_8_2_MSVC2022_64bit-Release

Maybe it would work if build dir is inside source dir? 
Most important that it works in the CI, not my system, though...
